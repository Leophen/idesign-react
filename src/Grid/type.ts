export interface GridProps {
  /**
   * 栅格每一项的垂直对齐方式
   * @default top
   */
  align?: 'top' | 'middle' | 'bottom';
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 栅格每一项间的间隔
   */
  gutter?: number;
  /**
   * 栅格水平排列方式
   * @default start
   */
  justify?: 'start' | 'end' | 'center' | 'space-around' | 'space-between';
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
}

export interface GridItemProps {
  /**
   * 栅格单项的垂直对齐方式
   */
  align?: 'top' | 'middle' | 'bottom';
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 栅格每一项间的间隔
   */
  gutter?: number;
  /**
   * 栅格左侧的间隔格数，间隔内不可以有栅格
   */
  offset?: number;
  /**
   * 栅格单项次序
   */
  order?: number;
  /**
   * 栅格占位格数，为 0 时相当于 display: none
   * @default 24
   */
  span?: number;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 自定义单项宽度
   */
  width?: React.CSSProperties["width"];
}
