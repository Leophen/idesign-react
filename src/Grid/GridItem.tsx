import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { GridItemProps } from './type';

const GridItem: React.FC<GridItemProps> = (props) => {
  const {
    align,
    children = '',
    className,
    gutter,
    order,
    offset,
    style,
    span = 24,
    width,
    ...restProps
  } = props;

  const limitNum = (num: number) => {
    let result = parseInt(num.toString());
    if (result > 24) {
      result = 24;
    }
    if (result < 0) {
      result = 0;
    }
    return result;
  };

  const computedWidth = (width: React.CSSProperties["width"]) => {
    if (isNaN(width as number)) {
      return width;
    }
    const result = width + 'px';
    return result;
  };

  return (
    <div
      className={classNames(
        'i-grid__item',
        `i-grid__item--span-${limitNum(span)}`,
        offset && `i-grid__item--offset-${limitNum(offset)}`,
        order && `i-grid__item--order-${limitNum(order)}`,
        align && `i-grid__item--align-${align}`,
        className,
      )}
      style={{
        ...(style || {}),
        ...({
          paddingLeft: gutter,
          paddingRight: gutter,
        } || {}),
        ...(width
          ? {
            flex: `0 0 ${computedWidth(width)}`,
            maxWidth: computedWidth(width),
          }
          : { flex: '1' }),
      }}
      {...restProps}
    >
      {children}
    </div>
  );
};

export default GridItem;
