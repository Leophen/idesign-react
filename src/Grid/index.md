---
nav:
  title: 组件
  path: /components
group:
  title: 基础组件
  order: 1
order: 4
---

# Grid 栅格

24 栅格系统是以规则的网格阵列来规范页面中的模块布局及信息分布，提高界面布局的一致性。

## 基本用法

使用 `Grid` 包裹栅格的每一项 `Grid.Item`。

`Grid.Item` 的 `span` 属性值使用 0（不显示）到 24（默认值）的值来表示其跨越范围。例如，三个等宽的列可以使用 `<Grid.Item span={8}>` 来创建。

如果一个 `Grid` 中 `Grid.Item` 的 `span` 总和超过 24，那么多余的 `Grid.Item` 会作为一个整体另起一行排列。

```tsx
import React from 'react';
import { Grid } from 'idesign-react';

const demoData = [
  Array(24).fill(1),
  Array(12).fill(2),
  Array(8).fill(3),
  Array(6).fill(4),
  Array(4).fill(6),
  Array(3).fill(8),
  Array(2).fill(12),
  Array(1).fill(24),
];

const App = () => {
  return (
    <>
      {/* 直接使用 */}
      <Grid>
        <Grid.Item span={8}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={8}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={8}>
          <div>GridItem3</div>
        </Grid.Item>
      </Grid>

      {/* 遍历使用 */}
      {demoData.map((grid, i) => (
        <Grid key={i}>
          {grid.map((item, j) => (
            <Grid.Item span={item} key={j}>
              <div>{item}</div>
            </Grid.Item>
          ))}
        </Grid>
      ))}

      {/* span 总和超出换行 */}
      <Grid>
        <Grid.Item span={8}>
          <div>span8</div>
        </Grid.Item>
        <Grid.Item span={8}>
          <div>span8</div>
        </Grid.Item>
        <Grid.Item span={12}>
          <div>span12</div>
        </Grid.Item>
      </Grid>
    </>
  );
};

export default App;
```

## 自定义单项宽度

`Grid.Item` 的 `width` 属性值可以设置单项固定宽度，`width` 为具体的宽度且优先级高于 `span`，而 `span` 设置的宽度为间隔格数（0 - 24）

```tsx
import React from 'react';
import { Grid } from 'idesign-react';

const App = () => {
  return (
    <>
      <Grid>
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item width={100}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item width="auto">
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4} width={120}>
          <div>GridItem4</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem5</div>
        </Grid.Item>
      </Grid>
    </>
  );
};

export default App;
```

## 区块间隔

`Grid` 的 `gutter` 属性值可以全局设置栅格中每一项的左右内边距。例如，`<Grid gutter={16}>` 将会为每一项 `Grid.Item` 添加左右内边距为 `8px`。

```tsx
import React from 'react';
import { Grid } from 'idesign-react';

const App = () => {
  return (
    <>
      <Grid gutter={0}>
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <Grid gutter={16}>
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <Grid gutter={32}>
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>
    </>
  );
};

export default App;
```

## 单项左侧间距

`Grid.Item` 的 `offset` 属性值可以单独设置该项的左侧外边距，单位为间隔格数。例如，`<Grid.Item offset={2}>` 将会向右偏移 2 个间隔的距离。

```tsx
import React from 'react';
import { Grid } from 'idesign-react';

const App = () => {
  return (
    <>
      <Grid>
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <Grid>
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4} offset={2}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <Grid gutter={16}>
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4} offset={2}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>
    </>
  );
};

export default App;
```

## 栅格水平排列方式

`Grid` 的 `justify` 属性值可以全局设置栅格水平的排版。

```tsx
import React from 'react';
import { Grid } from 'idesign-react';

const App = () => {
  return (
    <>
      <h3>start</h3>
      <Grid justify="start">
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <h3>center</h3>
      <Grid justify="center">
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <h3>end</h3>
      <Grid justify="end">
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <h3>space-between</h3>
      <Grid justify="space-between">
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <h3>space-around</h3>
      <Grid justify="space-around">
        <Grid.Item span={4}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>
    </>
  );
};

export default App;
```

## 栅格垂直对齐方式

`Grid` 的 `align` 属性值可以全局设置栅格中每一项的垂直对齐方式，而 `Grid.Item` 可通过 `align` 属性值单独设置该项的垂直对齐方式，优先级更高。

```tsx
import React from 'react';
import { Grid } from 'idesign-react';

const App = () => {
  return (
    <>
      <h3>top</h3>
      <Grid align="top">
        <Grid.Item span={4}>
          <div style={{ height: '80px' }}>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div style={{ height: '60px' }}>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <h3>middle</h3>
      <Grid align="middle">
        <Grid.Item span={4}>
          <div style={{ height: '80px' }}>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div style={{ height: '60px' }}>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <h3>bottom</h3>
      <Grid align="bottom">
        <Grid.Item span={4}>
          <div style={{ height: '80px' }}>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div style={{ height: '60px' }}>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>

      <h3>单项对齐设置</h3>
      <Grid align="bottom">
        <Grid.Item span={4}>
          <div style={{ height: '80px' }}>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={4} align="top">
          <div>AlignTop2</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div style={{ height: '60px' }}>GridItem3</div>
        </Grid.Item>
        <Grid.Item span={4}>
          <div>GridItem4</div>
        </Grid.Item>
      </Grid>
    </>
  );
};

export default App;
```

## 自定义单项排序

`Grid.Item` 的 `order` 属性值可以设置单项在整个栅格中的次序。

```tsx
import React from 'react';
import { Grid } from 'idesign-react';

const App = () => {
  return (
    <>
      <Grid gutter={10}>
        <Grid.Item span={2}>
          <div>1</div>
        </Grid.Item>
        <Grid.Item span={2}>
          <div>2</div>
        </Grid.Item>
        <Grid.Item span={2} order={4}>
          <div>3</div>
        </Grid.Item>
        <Grid.Item span={2} order={3}>
          <div>4</div>
        </Grid.Item>
        <Grid.Item span={2} order={3}>
          <div>5</div>
        </Grid.Item>
        <Grid.Item span={2} order={2}>
          <div>6</div>
        </Grid.Item>
        <Grid.Item span={2} order={1}>
          <div>7</div>
        </Grid.Item>
        <Grid.Item span={2} order={0}>
          <div>8</div>
        </Grid.Item>
        <Grid.Item span={2}>
          <div>9</div>
        </Grid.Item>
        <Grid.Item span={2}>
          <div>10</div>
        </Grid.Item>
      </Grid>
    </>
  );
};

export default App;
```

## Grid API

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| children | 内容 | `ReactNode` | `--` |
| className | 类名 | `string` | `--` |
| style | 自定义样式 | `CSSProperties` | `--` |
| justify | 栅格水平排列方式 | `"start"〡"end"〡"center"〡"space-around"〡"space-between"` | `"start"` |
| align | 栅格每一项的全局垂直对齐方式 | `"top"〡"middle"〡"bottom"` | `"top"` |
| gutter | 栅格每一项间的全局间隔 | `number` | `0` |

## GridItem API

| 属性      | 说明                   | 类型                        | 默认值  |
| --------- | ---------------------- | --------------------------- | ------- |
| children  | 内容                   | `ReactNode`                 | `--`    |
| className | 类名                   | `string`                    | `--`    |
| style     | 自定义样式             | `CSSProperties`             | `--`    |
| span      | 栅格宽度占比           | `number`                    | `24`    |
| width     | 栅格单项宽度值         | `number〡string`            | `--`    |
| align     | 栅格单项的垂直对齐方式 | `"top"〡"middle"〡"bottom"` | `"top"` |
| offset    | 栅格单项的左侧外边距   | `number`                    | `0`     |
| order     | 栅格单项次序           | `number`                    | `0`     |
