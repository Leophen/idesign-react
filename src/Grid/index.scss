$grid-list: (
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24
);

.i-grid {
  width: 100%;
  display: flex;
  flex-flow: row wrap;
  margin-right: 0px;
  margin-left: 0px;
}

.i-grid--align-top {
  align-items: flex-start;
}

.i-grid--align-middle {
  align-items: center;
}

.i-grid--align-bottom {
  align-items: flex-end;
}

.i-grid--justify-start {
  justify-content: flex-start;
}

.i-grid--justify-center {
  justify-content: center;
}

.i-grid--justify-end {
  justify-content: flex-end;
}

.i-grid--justify-space-between {
  justify-content: space-between;
}

.i-grid--justify-space-around {
  justify-content: space-around;
}

.i-grid__item {
  position: relative;
  max-width: 100%;
}

.i-grid__item--align-top {
  align-self: flex-start;
}

.i-grid__item--align-middle {
  align-self: center;
}

.i-grid__item--align-bottom {
  align-self: flex-end;
}

@each $item in $grid-list {
  .i-grid__item--span-#{$item} {
    @if $item == 0 { display: none; } @else { display: block; };
    max-width: calc((nth($item, 1) / 24) * 100%);
    flex: 0 0 calc((nth($item, 1) / 24) * 100%);
  }
  .i-grid__item--offset-#{$item} {
    margin-left: calc((nth($item, 1) / 24) * 100%);
  }
  .i-grid__item--order-#{$item} {
    order: $item;
  }
}
