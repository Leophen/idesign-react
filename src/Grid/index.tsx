import React from 'react';
import classNames from 'classnames';
import './index.scss';

export interface GridProps {
  /**
   * 栅格每一项的垂直对齐方式
   * @default top
   */
  align?: 'top' | 'middle' | 'bottom';
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 类名
   */
  className?: string;
  /**
   * 栅格每一项间的间隔
   */
  gutter?: number;
  /**
   * 栅格水平排列方式
   * @default start
   */
  justify?: 'start' | 'end' | 'center' | 'space-around' | 'space-between';
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
}

export interface GridItemProps {
  /**
   * 栅格单项的垂直对齐方式
   */
  align?: 'top' | 'middle' | 'bottom';
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 类名
   */
  className?: string;
  /**
   * 栅格每一项间的间隔
   */
  gutter?: number;
  /**
   * 栅格左侧的间隔格数，间隔内不可以有栅格
   */
  offset?: number;
  /**
   * 栅格单项次序
   */
  order?: number;
  /**
   * 栅格占位格数，为 0 时相当于 display: none
   * @default 24
   */
  span?: number;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 自定义单项宽度
   */
  width?: string | number;
}

const GridItem: React.FC<GridItemProps> = (props) => {
  const {
    align,
    children = '',
    className,
    gutter,
    order,
    offset,
    style,
    span = 24,
    width,
    ...others
  } = props;

  const limitNum = (num: number) => {
    let result = parseInt(num.toString());
    if (result > 24) {
      result = 24;
    }
    if (result < 0) {
      result = 0;
    }
    return result;
  };

  const computedWidth = (width: string | number) => {
    if (isNaN(width as number)) {
      return width;
    }
    const result = width + 'px';
    return result;
  };

  return (
    <div
      className={classNames(
        'i-grid__item',
        `i-grid__item--span-${limitNum(span)}`,
        offset && `i-grid__item--offset-${limitNum(offset)}`,
        order && `i-grid__item--order-${limitNum(order)}`,
        align && `i-grid__item--align-${align}`,
        className,
      )}
      style={{
        ...(style || {}),
        ...({
          paddingLeft: gutter,
          paddingRight: gutter,
        } || {}),
        ...(width
          ? {
              flex: `0 0 ${computedWidth(width)}`,
              maxWidth: computedWidth(width),
            }
          : { flex: '1' }),
      }}
      {...others}
    >
      {children}
    </div>
  );
};

const Grid: React.FC<GridProps> & { Item: React.ElementType } = (props) => {
  const { align, children = '', className, gutter, justify, style, ...others } = props;

  return (
    <div
      className={classNames(
        'i-grid',
        align && `i-grid--align-${align}`,
        justify && `i-grid--justify-${justify}`,
        className,
      )}
      style={{ ...style }}
      {...others}
    >
      {/* 将 gutter 传入 children */}
      {React.Children.map(children, (child) => {
        if (!React.isValidElement(child)) {
          return null;
        }
        const childProps = {
          gutter,
        };
        return React.cloneElement(child, childProps);
      })}
    </div>
  );
};

Grid.Item = GridItem;

Grid.displayName = 'Grid';
GridItem.displayName = 'GridItem';

export default Grid;
