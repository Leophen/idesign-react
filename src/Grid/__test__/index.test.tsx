import { render } from '@testing-library/react';
import Grid from '../index';

describe('Grid 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Grid>
        <Grid.Item span={8}>
          <div>GridItem1</div>
        </Grid.Item>
        <Grid.Item span={8}>
          <div>GridItem2</div>
        </Grid.Item>
        <Grid.Item span={8}>
          <div>GridItem3</div>
        </Grid.Item>
      </Grid>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

