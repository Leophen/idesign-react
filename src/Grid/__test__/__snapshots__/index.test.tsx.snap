// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-grid"
  >
    <div
      class="i-grid__item i-grid__item--span-8"
      style="flex: 1;"
    >
      <div>
        GridItem1
      </div>
    </div>
    <div
      class="i-grid__item i-grid__item--span-8"
      style="flex: 1;"
    >
      <div>
        GridItem2
      </div>
    </div>
    <div
      class="i-grid__item i-grid__item--span-8"
      style="flex: 1;"
    >
      <div>
        GridItem3
      </div>
    </div>
  </div>
</DocumentFragment>
`;
